### ISSUE TITLE
[AMP-STORY-PLAYER] Feature request : Control the player programmatically

### ISSUE THREAD

**Issue by edoudou** [ampproject/amphtml#27331]
## Description

We need to have a way to control the player programmatically. 

## Solution

An API for the web-component to allow basic controls over the player, including :

- The possibility to add stories to the player programmatically.
- The possibility to remove stories from the player programmatically.
- The possibility to change the current story being displayed by the player (with or without animation)
- The possibility to pause the content of the player (disabling auto-transition, pausing videos and audios) and to play it again programmatically.
- The possibility to mute the content of the player.

**Comment by newmuis** [comment]
@edoudou can you elaborate on the use case for `The possibility to remove stories from the player programmatically`?  This seems like it might cause some weird user experiences.

The others all sound good to me.

**Comment by edoudou** [comment]
@newmuis I added because it goes with `The possibility to add stories to the player programmatically`, but we don't have a real use case for it as of now. 
However, I could see it being used in a tool, where someone could build a story player and add and remove stories to see the result.
It could also be used if you want to allow someone to hide content once they read it.

**Comment by Enriqe** [comment]
### Adding stories to the player programmatically

#### addStory

*storyUrl* This method will take a URL pointing to a story

`ampStoryPlayer.addStory(storyUrl, opt_description);` 

Note that the player will try to load the story from the cache that was specified in the player level.



### Change the current story being displayed by the player 


The possibility to pause the content of the player (disabling auto-transition, pausing videos and audios) and to play it again programmatically.
The possibility to mute the content of the player.

**Comment by Enriqe** [comment]
Proposed API:

###  Adding stories to the player programmatically
* Method signature: `player.addStory(storyUrl, opt_description)`
* Note that the player will try to load the story from the cache that was specified at the player level. e.g. `cdn.ampproject.org` / `www.bing-amp.com`

#### Parameters:

* **storyUrl:** (string) URL of the story to add.
* **opt_description:** *optional* (string) Short description of the story.

###  Change the current story being displayed by the player
* Method signature: `player.showStory(storyUrl)`
* Story must have been previously added to the player. Otherwise an error will be thrown.

#### Parameters:

* **storyUrl:** (string) URL of the story to show.

###  Pause current story in the player
* Method signature: `player.pause()`

###  Play current story in the player
* Method signature: `player.play()`

### The possibility to mute the content of the player.
* Method signature: `player.mute()`

**Comment by newmuis** [comment]
nit: it might be interesting or useful to allow switching stories by index rather than URL.  For example, if a player renders pagination dots, if it can only switch by URL it will have to maintain a separate mapping of index to URL

**Comment by gmajoulet** [comment]
+1 for an API to navigate between stories, on top of the `.show()`, that is useful to re-use a single player within an app (ie: use one player to show the story user clicks on).
It could accept an index (might be hard for publishers to keep track of the current index though), or a delta (cf the [`history.go`](https://developer.mozilla.org/en-US/docs/Web/API/History/go) API).

> `addStory`

If it's meant to be used to create an infinite feed of stories, publishers would likely add stories in batches. What do you think of accepting an array of URLs instead?
What is the `description` parameter used for? 

> The possibility to mute the content of the player

Should we also add the `unmute` method? In a native context where the webview is properly configured to allow autoplaying with sound, this is needed to build an infinite feed of stories where the muted state is persisted from one story to the next one.

**Comment by Enriqe** [comment]
> it might be interesting or useful to allow switching stories by index rather than URL
> +1 for an API to navigate between stories, on top of the .show()

WDYT about having `loadUrl(url)` and `go(delta)`. 
* `loadUrl(url)` would load a story `url`, even if it hasn't been added to the player yet
* `go(delta)` navigates to a story inside the player, similar to the [history.go](https://developer.mozilla.org/en-US/docs/Web/API/History/go) API.

> If it's meant to be used to create an infinite feed of stories, publishers would likely add stories in batches. What do you think of accepting an array of URLs instead?

That's a good point. I updated `add()` to receive an array of `StoryObj` (see updated spec).

> Adding an option to animate or not the transition to this story, as suggested in the original feature request, also makes sense.

Must have missed it, added to the spec.

> What is the description parameter used for?

Since we'll be likely appending an `<a>` tag when adding the story, I think it makes sense to also add a `description` inside, to be consistent with the original player API (we recommend people to write a description in each `<a>` tag, which helps with SEO.

>  Should we also add the unmute method?

Woops, added it.

**Comment by gmajoulet** [comment]
That all mostly looks good to me! Thanks for getting this together!

A few nits:

> `loadUrl(url)`

So `load` would basically be `add` + `show`?
The idea sounds good. What do you think of having `show` automatically call `add` internally if it wasn't added before? This way, we can keep a super simple `add` and `show` API that should cover all use cases.
`add` to add stories to the player
`show` to select one story via its URL, which adds it if it wasn't present before
`go(delta)` to go navigate if you have custom controls like prev/next buttons, or pagination dots as Jon mentioned

> opt_description: optional (string) Short description of the story

Let's make it clear that it's the anchor's `title`, through its name and description?

**Comment by newmuis** [comment]
optional: Does it make sense to use `select` or `selectDocument` instead of `go` to match the actual viewer API?

**Comment by Enriqe** [comment]
> What do you think of having show automatically call add internally if it wasn't added before?

SG! Updated it.

> Let's make it clear that it's the anchor's title, through its name and description?
Done.

> optional: Does it make sense to use select or selectDocument

Did you mean using `selectDocument({'next': true})`? If we follow this route then we would miss out on using an index as the parameter, which would make things easier for pagination dots use case you mentioned..

**Comment by newmuis** [comment]
I just meant retaining the `selectDocument` name, and perhaps the object.  But yeah, maybe it's fine to not do that.

**Comment by gmajoulet** [comment]
Additionally, we should provide some method for publishers to know when they can start interacting with the API.

I'm not 100% sure of how this could work, but there are two possible directions:

1. We expose a method that either tells whether the player is ready to do mute/unmute/go/... operations, or returns a promise that resolves when the player is ready, or triggers a "ready" event...
2. We don't expose anything but store all the method calls and plays them in order once the first story is ready. Eg: story is loading but publisher calls `unmute`, which won't be called until the first story is loaded and viewer communication is set up

Maybe the best solution is a mix of (1) and (2) 🤷‍♂️

**Comment by gmajoulet** [comment]
As discussed last week, this API to interact looks good to me. We should also add some events for publishers to know about the player's state, and know when to interact with it. A few events I have in mind: new story is focused and its index, how many stories are remaining, audio got muted/unmuted, etc...

**Comment by Enriqe** [comment]
Continuing the discussion in #28416

**Comment by gmajoulet** [comment]
Tracked in https://github.com/ampproject/amphtml/issues/28416
